File: Cbor/CBORTag37.cs
/*
Written in 2014 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */
using System;
using PeterO;

namespace PeterO.Cbor {
    /// <summary>Description of CBORTag37.</summary>
  internal class CBORTag37 : ICBORTag, ICBORConverter<Guid>
  {
    /// <summary>Not documented yet.</summary>
    /// <returns>A CBORTypeFilter object.</returns>
    public CBORTypeFilter GetTypeFilter() {
      return CBORTypeFilter.ByteString;
    }

    /// <summary>Not documented yet.</summary>
    /// <param name='obj'>A CBORObject object. (2).</param>
    /// <returns>A CBORObject object.</returns>
    public CBORObject ValidateObject(CBORObject obj) {
      if (obj.Type != CBORType.ByteString) {
        throw new CBORException("UUID must be a byte string");
      }
      byte[] bytes = obj.GetByteString();
      if (bytes.Length != 16) {
        throw new CBORException("UUID must be 16 bytes long");
      }
      return obj;
    }

    internal static void AddConverter() {
      CBORObject.AddConverter(typeof(Guid), new CBORTag37());
    }

    /// <summary>Converts a UUID to a CBOR object.</summary>
    /// <param name='obj'>A UUID.</param>
    /// <returns>A CBORObject object.</returns>
    public CBORObject ToCBORObject(Guid obj) {
      byte[] bytes = obj.ToByteArray();
      byte[] bytes2 = new byte[16];
      Array.Copy(bytes, bytes2, 16);
      // Swap the bytes to conform with the UUID RFC
      bytes2[0] = bytes[3];
      bytes2[1] = bytes[2];
      bytes2[2] = bytes[1];
      bytes2[3] = bytes[0];
      bytes2[4] = bytes[5];
      bytes2[5] = bytes[4];
      bytes2[6] = bytes[7];
      bytes2[7] = bytes[6];
      return CBORObject.FromObjectAndTag(bytes2, (int)37);
    }
  }
}

File: src/main/java/com/upokecenter/cbor/CBORTag37.java
package com.upokecenter.cbor;
/*
Written in 2014 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */

import java.util.UUID;

    /**
     * Description of CBORTag37.
     */
  class CBORTag37 implements ICBORTag, ICBORConverter<UUID> {

    /**
     * Not documented yet.
     *
     * @return A CBORTypeFilter object.
     */
    public CBORTypeFilter GetTypeFilter() {
      return CBORTypeFilter.ByteString;
    }

    /**
     * {@inheritDoc}
     *
     * Not documented yet.
     */
    public CBORObject ValidateObject(final CBORObject obj) {
      if (obj.getType() != CBORType.ByteString) {
        throw new CBORException("UUID must be a byte string");
      }
      if (obj.GetByteString().length != 16) {
        throw new CBORException("UUID must be 16 bytes long");
      }
      return obj;
    }

    static void AddConverter() {
      CBORObject.AddConverter(UUID.class, new CBORTag37());
    }

    /**
     * <p>ToCBORObject.</p>
     *
     * @param obj a {@link java.util.UUID} object.
     * @return a {@link com.upokecenter.cbor.CBORObject} object.
     */
    public CBORObject ToCBORObject(final UUID obj) {
      byte[] bytes2 = new byte[16];
      long lsb = obj.getLeastSignificantBits();
      long msb = obj.getMostSignificantBits();
      bytes2[0] = (byte)((msb >> 56) & 0xFFL);
      bytes2[1] = (byte)((msb >> 48) & 0xFFL);
      bytes2[2] = (byte)((msb >> 40) & 0xFFL);
      bytes2[3] = (byte)((msb >> 32) & 0xFFL);
      bytes2[4] = (byte)((msb >> 24) & 0xFFL);
      bytes2[5] = (byte)((msb >> 16) & 0xFFL);
      bytes2[6] = (byte)((msb >> 8) & 0xFFL);
      bytes2[7] = (byte)((msb) & 0xFFL);
      bytes2[8] = (byte)((lsb >> 56) & 0xFFL);
      bytes2[9] = (byte)((lsb >> 48) & 0xFFL);
      bytes2[10] = (byte)((lsb >> 40) & 0xFFL);
      bytes2[11] = (byte)((lsb >> 32) & 0xFFL);
      bytes2[12] = (byte)((lsb >> 24) & 0xFFL);
      bytes2[13] = (byte)((lsb >> 16) & 0xFFL);
      bytes2[14] = (byte)((lsb >> 8) & 0xFFL);
      bytes2[15] = (byte)((lsb) & 0xFFL);
      return CBORObject.FromObjectAndTag(bytes2, 37);
    }

  }
